src/move/emojicoin_dot_fun/sources/emojicoin_dot_fun.move

```diff
@@ -240,13 +247,63 @@ module emojicoin_dot_fun::emojicoin_dot_fun {
         verified_bytes
     }
 
+    inline fun mul_div(
+        a: u64,
+        b: u64,
+        c: u64,
+    ): u128 {
+        (a as u128) * (b as u128) / (c as u128)
+    }
+
     inline fun assert_valid_coin_types<Emojicoin, EmojicoinLP>(market_address: address) {
         assert!(
             exists<LPCoinCapabilities<Emojicoin, EmojicoinLP>>(market_address),
             E_INVALID_COIN_TYPES
         );
     }
 
+    inline fun fdv_market_cap_start_end(
+        reserves_start: Reserves,
+        reserves_end: Reserves,
+        supply_minuend: u64,
+    ): (
+        u128, // FDV at start.
+        u128, // Market cap at start.
+        u128, // FDV at end.
+        u128, // Market cap at end.
+    ) {
+        let (fdv_start, market_cap_start) = fdv_market_cap(reserves_start, supply_minuend);
+        let (fdv_end, market_cap_end) = fdv_market_cap(reserves_end, supply_minuend);
+        (fdv_start, market_cap_start, fdv_end, market_cap_end)
+    }
+
+    inline fun fdv_market_cap(
+        reserves: Reserves,
+        supply_minuend: u64,
+    ): (
+        u128, // FDV.
+        u128, // Market cap.
+    ) {
+        let base = reserves.base;
+        let quote = reserves.quote;
+        (
+            mul_div(quote, EMOJICOIN_SUPPLY, base), // FDV.
+            mul_div(quote, supply_minuend - base, base), // Market cap.
+        )
+    }
+
+    inline fun assign_supply_minuend_reserves_ref_mut(
+        market_ref_mut: &mut Market,
+        starts_in_bonding_curve: bool
+    ): (
+        u64,
+        &mut Reserves,
+    ) {
+        if (starts_in_bonding_curve)
+            (BASE_VIRTUAL_CEILING, &mut market_ref_mut.clamm_virtual_reserves) else
+            (EMOJICOIN_SUPPLY, &mut market_ref_mut.cpamm_real_reserves)
```

Comment:
```suggestion
        if (starts_in_bonding_curve) {
            (BASE_VIRTUAL_CEILING, &mut market_ref_mut.clamm_virtual_reserves)
        } else {
            (EMOJICOIN_SUPPLY, &mut market_ref_mut.cpamm_real_reserves)
        }
```

This is a bit of nit but any way we can do the brace if/else syntax? It's a lot easier to parse at a glance, I missed the else block here for a few seconds 😅